util/oteltest/src/oteltest/private.py

```diff
@@ -0,0 +1,283 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import glob
+import importlib
+import inspect
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+import time
+import typing
+import venv
+from pathlib import Path
+
+from google.protobuf.json_format import MessageToDict
+from oteltest import OtelTest, telemetry
+from oteltest.sink import GrpcSink, RequestHandler
+
+from opentelemetry.proto.collector.logs.v1.logs_service_pb2 import (
+    ExportLogsServiceRequest,
+)
+from opentelemetry.proto.collector.metrics.v1.metrics_service_pb2 import (
+    ExportMetricsServiceRequest,
+)
+from opentelemetry.proto.collector.trace.v1.trace_service_pb2 import (
+    ExportTraceServiceRequest,
+)
+
+
+def run(script_path: str, wheel_file: str, venv_parent_dir: str):
+    temp_dir = venv_parent_dir or tempfile.mkdtemp()
+    print(f"- Using temp dir for venvs: {temp_dir}")
+
+    if os.path.isdir(script_path):
+        handle_dir(script_path, temp_dir, wheel_file)
+    elif os.path.isfile(script_path):
+        handle_file(script_path, temp_dir, wheel_file)
+    else:
+        print(f"- {script_path} does not exist")
+        return
+
+
+def handle_dir(script_path, temp_dir, wheel_file):
+    sys.path.append(script_path)
+    for script in ls_scripts(script_path):
+        print(f"- Setting up environment for dir {script}")
+        setup_script_environment(temp_dir, script_path, script, wheel_file)
+
+
+def handle_file(script_path, temp_dir, wheel_file):
+    print(f"- Setting up environment for file {script_path}")
+    script_dir = os.path.dirname(script_path)
+    sys.path.append(script_dir)
+    setup_script_environment(
+        temp_dir, script_dir, os.path.basename(script_path), wheel_file
+    )
+
+
+def ls_scripts(script_dir):
+    original_dir = os.getcwd()
+    os.chdir(script_dir)
+    scripts = glob.glob("*.py")
+    os.chdir(original_dir)
+    return scripts
+
+
+def setup_script_environment(venv_parent, script_dir, script, wheel_file):
+    handler = AccumulatingHandler()
+    sink = GrpcSink(handler)
+    sink.start()
+
+    module_name = script[:-3]
+    oteltest_instance: OtelTest = load_test_class_for_script(module_name)()
+
+    script_venv = Venv(str(Path(venv_parent) / module_name))
+    script_venv.create()
+
+    pip_path = script_venv.path_to_executable("pip")
+
+    run_subprocess([pip_path, "install", (wheel_file or "oteltest")])
```

Comment:
oteltest version should probably be pinned to the same of the installed outside

Reply:
Excellent point. Will fix.

Reply:
This also should be installed after I've installed my requirements since they may have stricter dependencies on opentelemetry packages versions

Reply:
BTW, there's no reason for us to require that an oteltest script define a class that formally implements the OtelTest base class. Longer term, I intend for the OtelTest base class to be just a convenience so your IDE can fill in the missing methods for you. But if you just have a class called MyOtelTest that doesn't inherit from anything, but it does implement all of the OtelTest methods, I think we could easily consider that a suitable implementation. I'd like to not require that target scripts depend on oteltest at all.

Reply:
(And we could allow methods to be missing as well)

Reply:
Adding an `"OtelTest" in value.__name__` value will work for me